src/backend/app/projects/project_crud.py

```diff
@@ -13,56 +12,115 @@
     """
-    new_project_id = await db.execute(query)
+    new_project_id = await db.execute(query, values = {
+        "author_id": 1,
+        "name": project_metadata.name,
+        "short_description": project_metadata.short_description,
+        "description": project_metadata.description,
+        "per_task_instructions": project_metadata.per_task_instructions,
+        "status": "DRAFT",
+        "visibility": "PUBLIC",
+        "mapper_level": "INTERMEDIATE",
+        "priority": "MEDIUM",
+        "outline": str(project_metadata.outline),
+    })
 
     if not new_project_id:
         raise HTTPException(status_code=500, detail="Project could not be created")
     # Fetch the newly created project using the returned ID
-    select_query = f"""
+    select_query = """
         SELECT id, name, short_description, description, per_task_instructions, outline
         FROM projects
-        WHERE id = '{new_project_id}'
+        WHERE id = :new_project_id
     """
-    new_project = await db.fetch_one(query=select_query)
+    
+    new_project = await db.fetch_one(select_query,{"new_project_id": new_project_id})
     return new_project
 
 
 async def get_project_by_id(
-    db: Session = Depends(database.get_db), project_id: Optional[int] = None
-) -> db_models.DbProject:
-    """Get a single project by id."""
-    db_project = (
-        db.query(db_models.DbProject)
-        .filter(db_models.DbProject.id == project_id)
-        .first()
-    )
-    return await convert_to_app_project(db_project)
+    db: Database = Depends(database.encode_db), project_id: Optional[int] = None
+):
+    """Get a single project &  all associated tasks by ID."""
+    # check the project in Database
+    raw_sql = """
+    SELECT
+        projects.id,
+        projects.short_description,
+        projects.description,
+        projects.author_id,
+        users.name AS author_name
+    FROM projects
+    JOIN users ON projects.author_id = users.id
+    WHERE projects.id = :project_id
+    LIMIT 1;
+    """
+
+    project_record = await db.fetch_one(raw_sql, {"project_id": project_id})
+    if not project_record:
+        raise HTTPException(
+            status_code=HTTPStatus.BAD_REQUEST, detail="Project not found."
+        )
+
+    # get tasks of project
+    query = """ SELECT * from tasks WHERE project_id = :project_id;"""
+    task_records = await db.fetch_all(query, {"project_id": project_id})
+    tasks = await convert_to_app_tasks(task_records)
+    # Count the total tasks
+    task_count = len(tasks)
+    return {
+        "project": project_record,
+        "task_count": task_count,
+        "tasks": tasks,
+    }
+
+
+async def convert_to_app_tasks(
+    db_tasks: List[db_models.DbTask],
+) -> List[project_schemas.TaskOut]:
+    """Legacy function to convert db models --> Pydantic.
+
+    TODO refactor to use Pydantic model methods instead.
+    """
+    if db_tasks and len(db_tasks) > 0:
+
+        async def convert_task(task):
```

Comment:
So this kind of thing is also not necessary.

The way FastAPI and Pydantic work together is like this:
- Define your schemas in Pydantic
- Combine together the schemas however you need to make a model that is returned from an endpoint
- Specify the `response_model` on the endpoint.
- FastAPI / Pydantic handle the validation and serialisation for you.

So say you had a task model:

```python
TaskOut(BaseModel):
    something: str
    something2: int
```

then you can use this in your Project model:

```python
ProjectOut(BaseModel):
    id: int
    tasks: list[TaskOut]
```

Then you can use that in your endpoint:

```python
@app.get("/projects", response_model=list[ProjectOut])
...
    # logic here that queries the db and returns the db data, including associated tasks
    return db_projects
```

- Then when the endpoint is called, the projects are returned from the db.
- FastAPI will use Pydantic to serialise the db data into the format specified in the model (i.e. omits fields it doesn't need and includes the fields it does).
- FMTM has some examples of this that could help 👍

Reply:
```
@router.get("/{project_id}", tags=["Projects"], response_model=project_schemas.ProjectOut)
async def read_project(
    project_id: int,
    db: Database = Depends(database.encode_db),
):
    """Get a specific project and all associated tasks by ID."""
    project = await project_crud.get_project_by_id(db, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return project
```
```
async def get_project_by_id(
    db: Database = Depends(database.encode_db), project_id: Optional[int] = None
):
    """Get a single project &  all associated tasks by ID."""
    # check the project in Database
    raw_sql = """
    SELECT
        projects.id,
        projects.name,
        projects.short_description,
        projects.description,
        projects.per_task_instructions,
        projects.author_id,
        projects.outline,
        users.name AS author_name
    FROM projects
    JOIN users ON projects.author_id = users.id
    WHERE projects.id = :project_id
    LIMIT 1;
    """

    project_record = await db.fetch_one(raw_sql, {"project_id": project_id})
    # Fetch associated tasks
    query = """ SELECT * from tasks WHERE project_id = :project_id;"""
    task_records = await db.fetch_all(query, {"project_id": project_id})
   
    return  {
        "id": project_record["id"],
        "name": project_record["name"],
        "short_description": project_record["short_description"],
        "description": project_record["description"],
        "per_task_instructions": project_record["per_task_instructions"],
        "outline": project_record["outline"],
        "tasks": task_records,
        "task_count": len(task_records)
    }
```
```
class TaskOut(BaseModel):
    """Base project model."""

    id: int
    project_task_index: int
    project_task_name: Optional[str] = None
    outline: Any = Field(exclude=True)

    @computed_field
    @property
    def outline_geojson(self) -> Optional[Feature]:
        """Compute the geojson outline from WKBElement outline."""
        if not self.outline:
            return None
        wkb_data = bytes.fromhex(self.outline)
        geom = wkb.loads(wkb_data)
        bbox = geom.bounds  # Calculate bounding box
        return str_to_geojson(self.outline, {"id": self.id, "bbox": bbox}, self.id)

```
```
class ProjectOut(BaseModel):
    """Base project model."""

    id: int
    name: str
    short_description: str
    description: str
    per_task_instructions: Optional[str] = None
    outline: Any = Field(exclude=True)
    tasks: list[TaskOut] = None
    task_count: int = None

    @computed_field
    @property
    def outline_geojson(self) -> Optional[Feature]:
        """Compute the geojson outline from WKBElement outline."""
        if not self.outline:
            return None
        wkb_data = bytes.fromhex(self.outline)
        geom = wkb.loads(wkb_data)
        bbox = geom.bounds  # Calculate bounding box
        return str_to_geojson(self.outline, {"id": self.id, "bbox": bbox}, self.id)
```
@spwoodcock  I've been working on implementing project details along with their associated tasks and counting the total tasks. As you can see above, I'm using raw SQL to fetch projects and their associated tasks. While this approach aligns with typical FastAPI and Pydantic usage patterns, I feel somewhat uncomfortable with the structure of the get_project_by_id function's return statement:
 ```
{
    "id": project_record["id"],
    "name": project_record["name"],
    "short_description": project_record["short_description"],
    "description": project_record["description"],
    "per_task_instructions": project_record["per_task_instructions"],
    "outline": project_record["outline"],
    "tasks": task_records,
    "task_count": len(task_records)
}
```
I'm unsure if this is the best approach. Could you provide guidance or suggest an alternative method to transform raw data into a format that validates against a Pydantic schema?

Reply:
This is lots better - great job!

You are right the current returned dict is prone to breaking if fields change.
You could simply:

```python
return await db.fetch_all(query, {"project_id": project_id}) 
```

This will pass the db object through the pydantic schema, filtering out fields not required and running the validators / serializers to JSON for the endpoint response 😄

FastAPI / Pydantic are neat